doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,214 @@
+
+Note that there is intentionally no way to view the contents of secrets.
+
+### System catalog changes
+
+A new `mz_secrets` table with the following structure will describe the
+available secrets in the system:
+
+Field       | Type      | Meaning
+------------|-----------|--------
+`id`        | `bigint`  | The ID of the secret.
+`schema_id` | `bigint`  | The ID of the schema to which the secret belongs.
+`name`      | `text`    | The name of the secret.
+
+Note that, again, the contents of the secrets are not exposed.
+
+### Referencing secrets
+
+There is presently no way to use a secret. The existing `CREATE SOURCE` and
+`CREATE SINK` commands will be extended to support referencing secrets, but
+that work is left to a future design document.
+
+## Implementation
+
+### Catalog changes
+
+The [`CatalogItem`] enum will be extended with an additional `Secret` variant:
+
+```rust
+pub enum CatalogItem {
+    // ...
+    Secret,
+}
+```
+
+Unlike the other catalog items, there is no data associated with a secret. The
+name and ID of the secret are stored in the `CatalogEntry` struct, and the
+actual data for the secret will be stored in a secret backend (see next
+section). Therefore there is presently no additional metadata that needs to be
+recorded about a secret.
+
+When serializing a secret, its `create_sql` will be synthesized as:
+
+```sql
+CREATE SECRET <name> AS '******'
+```
+
+When deserializing a secret on bootstrap, the `'******'` will be ignored.
+
+### Secrets controller
+
+The `Coordinator` will host a `SecretsController` that securely handles the
+storage of secrets. It will interact with the controller via the following
+interface.
+
+```rust
+/// Securely stores secrets.
+pub trait SecretsController {
+    /// Applies the specified secret operations in bulk.
+    ///
+    /// Implementations must apply the operations atomically. If the method
+    /// returns `Ok(())`, then all operations have been applied successfully;
+    /// if the method returns `Err(())`, then none of the operations have been
+    /// applied.
+    ///
+    /// Implementations are permitted to reject combinations of operations which
+    /// they cannot apply atomically.
+    fn apply(&mut self, ops: Vec<SecretOp>) -> Result<(), anyhow::Error>;
+}
+
+/// An operation on a [`SecretsController`].
+pub enum SecretOp {
+    /// Create or update the contents of a secret.
+    Ensure {
+        /// The ID of the secret to create or update.
+        id: GlobalId,
+        /// The binary contents of the secret.
+        contents: Vec<u8>,
+    },
+    /// Delete a secret.
+    Delete {
+        /// The id of the secret to delete.
+        id: GlobalId
+    }
+}
+```
+
+There will be two concrete implementations of the `SecretsController`:
+
+  * The `FilesystemSecretsController` will store secrets on the local filesystem.
+    It will not support multi-operation atomicity, but will support atomically
+    updating or removing one secret at a time.
```

Comment:
for my edification, when would `FilesystemSecretsController` get used? running things locally / single-node deployments?

Reply:
Yes, exactly!